### PR TITLE
Implement fallback mode

### PR DESCRIPTION
**Pull request by jeremydstone**
This PR is a revision to my earlier, unsubmitted fallback mode PR (https://github.com/code-dot-org/code-dot-org/pull/18826). That branch got in a bad state so I am doing revisions in this new branch, but will leave that PR open for reference.

This PR addresses feedback that the milestone post behavior had gotten too confusing. We had gone from two booleans (postMilestone and postFinalMilestone) to three (postFailedRunMilestone). 

In turned out that we have never actually altered the value of of postFinalMilestone to anything other than true, did not have any UI to do that, nor forsee any condition where we might ever do that in the future. So I removed it. Milestone posts on the final level of a stage now always happen and there is no way to turn them off. That leaves two booleans (postMilestone and postFailedRunMilestone) which have three allowable combinations. Ideally this would be a three-value enum rather than two booleans. It turns out the Gatekeeper API fundamentally wants to deal with booleans (the main method `allows` returns a boolean) so we need to keep this expressed as two booleans at the Gatekeeper level. But for purposes of the client, we now convert the Gatekeeper booleans into to a three-value enum (POST_MILESTONE_MODE) and transmit that to the client. This reduces the surface area of the code that needs to understand the two boolean settings and makes the milestone post logic easier to read where the client posts it.

### CODE REVIEW

File: apps/src/code-studio/reporting.js [hunk 1]
```diff
@@ -244,9 +245,26 @@ reporting.sendReport = function (report) {
   // Post milestone iff the server tells us.
   // Check a second switch if we passed the last level of the script.
   // Keep this logic in sync with ActivitiesController#milestone on the server.
-  if (appOptions.postMilestone ||
-    (appOptions.postFinalMilestone && report.pass && appOptions.level.final_level)) {
+  var postMilestoneMode = appOptions.postMilestoneMode;
+  var postMilestone;
```
Comment (Bjvanminnen) [inline apps/src/code-studio/reporting.js]: nit: const/let instead of var

File: dashboard/app/controllers/activities_controller.rb [hunk 2]
```diff
@@ -34,9 +34,9 @@ def milestone
     # Check a second switch if we passed the last level of the script.
     # Keep this logic in sync with code-studio/reporting#sendReport on the client.
     post_milestone = Gatekeeper.allows('postMilestone', where: {script_name: script_name}, default: true)
-    post_final_milestone = Gatekeeper.allows('postFinalMilestone', where: {script_name: script_name}, default: true)
-    solved_final_level = solved && @script_level.try(:final_level?)
-    unless post_milestone || (post_final_milestone && solved_final_level)
+    post_failed_run_milestone = Gatekeeper.allows('postFailedRunMilestone', where: {script_name: script_name}, default: true)
+    final_level = @script_level.try(:final_level?)
+    unless (post_milestone && (post_failed_run_milestone || solved)) || final_level
```
Comment (Bjvanminnen) [inline dashboard/app/controllers/activities_controller.rb]: Worth a comment explaining that on the final_level we post_milestone on failure even if `postFailedRunMilestone` is false?

File: apps/src/code-studio/reporting.js (hunk 1 above)
Comment (jeremydstone) [reply apps/src/code-studio/reporting.js]: Thanks, updated.

File: dashboard/app/controllers/activities_controller.rb (hunk 2 above)
Comment (jeremydstone) [reply dashboard/app/controllers/activities_controller.rb]: Thx, added comment

**Comment by jeremydstone** [comment]
Received an LGTM from Brent over slack

**Pull request merged**